syntax = "proto3";

package message.definition.swift.mt;

message SwiftMtMessageDefinition {
    string uri = 1;
    repeated ObjDef objs = 5;
}

message ObjDef {
    oneof obj_one_of {
        SeqDef seq = 1;
        FldDef fld = 5;
    }
}

message SeqDef {
    string tag = 1;
    string value = 5;
    bool repeated = 10;
    bool mandatory = 20;

    repeated ObjDef children = 25;
}

message CmpDef {
    string name = 1;
    string constant = 5;
    string format = 10;
}

message OptnFrmt {
    string option = 1;
    string format = 5;
    repeated CmpDef components = 10;
}

message FldQlfr {
    string qualifier = 1;
    repeated string options = 5;
    string description = 10;
}

message FldDef {
    string tag = 1;
    string name = 5;
    repeated OptnFrmt format = 10;
    repeated FldQlfr qualifier = 15;
    bool mandatory = 20;
    bool repeated = 25;
}

message ComponentFormat {
    enum CharType {
        NUMERIC = 0;
        ALPHA = 1;
        ALPHA_NUMERIC = 2;
        HEX = 3;
        CHARSET_X = 4;
        CHARSET_Y = 5;
        CHARSET_Z = 6;
        DECIMAL = 7;
        BLANK = 8;
    }

    int32 min_length = 1;
    int32 max_length = 2;
    bool fixed_length = 3;
    int32 line_count = 4;
    CharType type = 5;
}